dspy/modeling/backends/text.py

```diff
@@ -0,0 +1,226 @@
+import logging
+import typing as t
+
+import regex
+from pydantic import Field
+
+from dspy.modeling.backends.base import BaseBackend
+from dspy.primitives import Completions, Example
+from dspy.signatures.signature import Signature, SignatureMeta
+
+logger = logging.getLogger(__name__)
+
+litellm_logger = logging.getLogger("LiteLLM")
+litellm_logger.setLevel(logging.WARNING)
+
+try:
+    from litellm import completion
+
+    _missing_litellm = False
+except ImportError:
+    _missing_litellm = True
+
+
+def passages_to_text(passages: t.Iterable[str]) -> str:
+    passages = list(passages)
+    if len(passages) == 0:
+        raise ValueError("Empty passages passed to format handler.")
+
+    if len(passages) == 1:
+        return passages[0]
+
+    return "\n".join(
+        [f"[{idx + 1}] <<{text}>>" for idx, text in enumerate(passages)],
+    )
+
+
+def format_answers(answers: t.Iterable[str]) -> str:
+    answers = list(answers)
+    if len(answers) == 0:
+        raise ValueError("Empty answers passed to format handler.")
+
+    return answers[0].strip()
+
+
+def default_format_handler(x: str) -> str:
+    if not isinstance(x, str):
+        raise ValueError(f"Wrong type passed to format handlers: {type(x)} should be a str")
+
+    return " ".join(x.split())
+
+
+DEFAULT_FORMAT_HANDLERS = {
+    "context": passages_to_text,
+    "passages": passages_to_text,
+    "answers": format_answers,
+}
+
+
+class TextBackend(BaseBackend):
+    """TextBackend takes a signature, its params, and predicts structured outputs leveraging LiteLLM."""
+
+    STANDARD_PARAMS: dict[str, t.Any] = {
+        "temperature": 0.5,
+        "max_tokens": 500,
+        "top_p": 1,
+        "frequency_penalty": 0,
+        "presence_penalty": 0,
+        "num_retries": 3,
+    }
+
+    model: str
+    default_params: dict[str, t.Any] = Field(default_factory=dict)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if _missing_litellm:
+            raise ImportError(
```

Comment:
i think we can add this to the try/except block like down for other LMs?

Reply:
and mention `pip install dspy-ai[litellm]` since it's an extra

Reply:
@arnavsinghvi11 @KCaverly we should make litellm a required dep probably

Reply:
what does litellm involve/add? how heavy?

Reply:
@okhat it's the backend that lets us have access to all the models from OpenAI, Anthropic, Huggingface, Cohere, etc. without having to build individual adapters for each one